src/run.cpp

```diff
@@ -0,0 +1,192 @@
+#include <stdio.h>
+#include <iostream>
+#include <set>
+
+#include "queue.hh"
+#include "run.hh"
+
+#define BUFSZ 512
+
+NullLogger* logger = nullptr;
+static PermutationMap* g_pmap = nullptr;
```

Comment:
Can we avoid globals here?

Reply:
I removed the globals, with the magic of C++ references!